### PR TITLE
[TACHYON-1332] Add tool for reading protocol buffer journal entries.

### PR DESCRIPTION
**Pull request by aaudiber**
https://tachyon.atlassian.net/browse/TACHYON-1332

### CODE REVIEW

File: servers/src/main/java/tachyon/master/journal/JournalTool.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ * <pre>
+ * java -cp assembly/target/tachyon-assemblies-0.9.0-SNAPSHOT-jar-with-dependencies.jar \
+ *   tachyon.master.journal.JournalTool journal/FileSystemMaster/log.out
+ * </pre>
+ */
+public final class JournalTool {
+  /** Number of dashes to place on a newline at the end of each journal entry. */
+  private static final int NUM_DASHES = 80;
+  /** Amount of time to wait before giving up on the user supplying a journal log via stdin. */
+  private static final long TIMEOUT_MS = 500;
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: Maybe this should be longer, especially if a user is trying to stream from a remote journal file?

File: servers/src/main/java/tachyon/master/journal/JournalTool.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+  public static void main(String[] args) throws FileNotFoundException, IOException {
+    InputStream inStream = null;
+    if (args.length == 0 && stdinHasData()) {
+      inStream = System.in;
+    } else if (args.length == 1) {
+      inStream = new FileInputStream(args[0]);
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: What happens if the argument is in a UFS, like s3 or hdfs?

File: servers/src/main/java/tachyon/master/journal/JournalTool.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    } else {
+      usage();
+      System.exit(-1);
+    }
+    JournalFormatter formatter = new ProtoBufJournalFormatter();
+    JournalInputStream journalStream = formatter.deserialize(inStream);
+    JournalEntry entry;
+    while ((entry = journalStream.getNextEntry()) != null) {
+      System.out.print(entry);
+      System.out.println(StringUtils.repeat('-', NUM_DASHES));
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: Can the dash delimiter string be saved instead of created each time?

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Thanks for this important tool! Could you also provide a simple example of an output?

File: servers/src/main/java/tachyon/master/journal/JournalTool.java (hunk 1 above)
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: Or make this configurable?

File: servers/src/main/java/tachyon/master/journal/JournalTool.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    while (System.in.available() == 0) {
+      if (System.currentTimeMillis() - start > TIMEOUT_MS) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static void usage() {
+    System.out.println(
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: What do you think of adding this tool to a tachyon shell command?

**Comment by yupeng9** [comment]
Thanks for adding this important tool!

File: servers/src/main/java/tachyon/master/journal/JournalTool.java (hunk 4 above)
Comment (aaudiber) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: I think that's a really good idea - usability goes way up when you don't have to invoke the giant java command. Done.

**Comment by aaudiber** [comment]
@gpang here's an example of running it:
```
➜  tachyon git:(JournalTool1) bin/tachyon readJournal < journal/FileSystemMaster/log.out
sequence_number: 842
inode_directory_id_generator {
  container_id: 0
  sequence_number: 11
}
--------------------------------------------------------------------------------
sequence_number: 843
inode_last_modification_time {
  id: 0
  last_modification_time_ms: 1448919684223
}
--------------------------------------------------------------------------------
sequence_number: 844
inode_file {
  id: 2684354559
  parent_id: 0
  name: "tmp6"
  persisted: false
  pinned: false
  creation_time_ms: 1448919684223
  last_modification_time_ms: 1448919684223
  block_size_bytes: 536870912
  length: 0
  completed: false
  cacheable: true
  ttl: -1
  user_name: ""
  group_name: ""
  permission: 0
}
--------------------------------------------------------------------------------
sequence_number: 845
complete_file {
  block_ids: 2667577344
  id: 2684354559
  length: 80
  op_time_ms: 1448919684322
}
--------------------------------------------------------------------------------
sequence_number: 846
inode_directory_id_generator {
  container_id: 0
  sequence_number: 11
}
--------------------------------------------------------------------------------
sequence_number: 847
inode_last_modification_time {
  id: 0
  last_modification_time_ms: 1448920023023
}
--------------------------------------------------------------------------------
sequence_number: 848
inode_file {
  id: 2701131775
  parent_id: 0
  name: "tmp9"
  persisted: false
  pinned: false
  creation_time_ms: 1448920023023
  last_modification_time_ms: 1448920023023
  block_size_bytes: 536870912
  length: 0
  completed: false
  cacheable: true
  ttl: -1
  user_name: ""
  group_name: ""
  permission: 0
}
--------------------------------------------------------------------------------
sequence_number: 849
complete_file {
  block_ids: 2684354560
  id: 2701131775
  length: 80
  op_time_ms: 1448920023082
}
--------------------------------------------------------------------------------
```

File: servers/src/main/java/tachyon/master/journal/JournalTool.java (hunk 1 above)
Comment (aaudiber) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: Made it configurable and defaulted to 5 seconds

File: servers/src/main/java/tachyon/master/journal/JournalTool.java (hunk 2 above)
Comment (aaudiber) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: Good point, that will usually be the case, so I've removed support for specifying the file on the command line. It was complicating the code and it isn't hard to just insert a "<" before the file argument.

File: servers/src/main/java/tachyon/master/journal/JournalTool.java (hunk 3 above)
Comment (aaudiber) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: Done

**Comment by aaudiber** [comment]
Jenkins, test this please.

File: servers/src/main/java/tachyon/master/journal/JournalTool.java (hunk 2 above)
Comment (gpang) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: Ok. What would the command look like for a file from hdfs or s3? Would we have to transfer it locally first, and then use this tool?

**Comment by gpang** [comment]
Awesome output! I just left a minor question. Thanks!

File: servers/src/main/java/tachyon/master/journal/JournalTool.java (hunk 2 above)
Comment (aaudiber) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: To use the command with hdfs or s3, you just need a way to send the contents of your journal file to stdout. I tested with hdfs by running `bin/hdfs dfs -cat  /user/aja/checkpoint.data | ~/Developer/tachyon/bin/tachyon readJournal`, and something similar could be done with s3 using `aws s3 cp` (http://stackoverflow.com/questions/28330907/how-to-use-aws-s3-cli-to-dump-files-to-stdout-in-bash).

**Comment by aaudiber** [comment]
@yupeng9 Which docs are you thinking of? The only documentation I see for `bin/tachyon` is the usage info for the script.

**Comment by yupeng9** [comment]
Sorry, I meant the project website like this http://tachyon-project.org/documentation/v0.8.2/Command-Line-Interface.html

**Comment by aaudiber** [comment]
How about we create a section under Dev Resources which describes the commands in bin/tachyon which are meant to be dev-facing? I think docs for the journal reader would fit better there. I've added https://tachyon.atlassian.net/browse/TACHYON-1379

**Comment by yupeng9** [comment]
That sounds good to me! Thanks.

**Comment by yupeng9** [comment]
LGTM

File: common/src/main/java/tachyon/Constants.java [hunk 5]
```diff
@@ -166,6 +166,8 @@
       "tachyon.master.journal.tailer.shutdown.quiet.wait.time.ms";
   public static final String MASTER_JOURNAL_TAILER_SLEEP_TIME_MS =
       "tachyon.master.journal.tailer.sleep.time.ms";
+  public static final String MASTER_JOURNAL_TOOL_STDIN_TIMEOUT_MS =
+      "tachyon.master.journal.tool.stdin.timeout.ms";
```
Comment (calvinjia) [inline common/src/main/java/tachyon/Constants.java]: Should this be a configuration value or just a parameter passed into the `readJournal` command?

Comment (aaudiber) [inline common/src/main/java/tachyon/Constants.java]: parameter makes more sense, I'll change it

File: servers/src/main/java/tachyon/master/journal/JournalTool.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ * </pre>
+ */
+public final class JournalTool {
+  /** Separator to place at the end of each journal entry. */
+  private static final String ENTRY_SEPARATOR = StringUtils.repeat('-', 80);
+  /** Amount of time to wait before giving up on the user supplying a journal log via stdin. */
+  private static final long TIMEOUT_MS =
+      new TachyonConf().getLong(Constants.MASTER_JOURNAL_TOOL_STDIN_TIMEOUT_MS);
+
+  public static void main(String[] args) throws FileNotFoundException, IOException {
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: Which call will throw `FileNotFoundException`?

Comment (aaudiber) [inline servers/src/main/java/tachyon/master/journal/JournalTool.java]: Never. Removed it, thanks

**Comment by calvinjia** [comment]
LGTM

**Pull request merged**